### PR TITLE
use SmallVec for InstantiatedPredicates lists

### PR DESCRIPTION
**Pull request by Nathan-Fenner**
I ran some basic counts for how many times `.instantiate_into()` or `.instantiate_into_identity()` were called to produce a `InstantiatedPredicates` struct during the `python x.py build` process (starting from a `clean` build).

I got these numbers:

```
total calls:     2_773_010
--------------------------
    identity:      525_181
    subst:       2_247_829
```

So this function is "reasonably" hot, with about 18% of calls being to the `identity` version.

I also counted how large each result was, summarized in the following table:

| Size | Percentage |
|:-:|:-:|
|<= 0|41.4%|
|<= 1|64.2%|
|<= 2|87.2%|
|<= 3|93.6%|
|<= 4|96.2%|
|<= 5|97.2%|
|<= 6|98.2%|
|<= 7|98.5%|
|<= 8|98.9%|
|<= 9|99.1%|

So, I figured that a `smallvec::SmallVec<[Predicate<'tcx>; 3]>` may be better suited to storing this value instead of a plain `Vec<Predicate<'tcx>>`. The value of ~~4~~ 3 was chosen mostly arbitrarily - although it makes the `InstantiatedPredicates` struct larger, my cursory inspection of how this type is used suggest it mostly lives as a temporary on the stack, where I expect that this will cause fewer problems.

---

An attempt at size 4 [only slightly improved some secondary metrics](https://github.com/rust-lang/rust/pull/109866#issuecomment-1493491618) some benchmarks while causing some regressions. 

---

Separately, to minimize the overall diff, I added new impls that make `SmallVec<[T; N]>` `TypeFoldable` and `TypeVisitable` just like `Vec<T>`; this makes it possible to `#[derive(TypeFoldable, TypeVisitable)]` in the new definition of `InstantiatedPredicates`.

```rs
impl<const N: usize, I: Interner, T: TypeFoldable<I>> TypeFoldable<I> for smallvec::SmallVec<[T; N]> {
    fn try_fold_with<F: FallibleTypeFolder<I>>(self, folder: &mut F) -> Result<Self, F::Error> {
        self.into_iter().map(|t| t.try_fold_with(folder)).collect()
    }
}

impl<const N: usize, I: Interner, T: TypeVisitable<I>> TypeVisitable<I> for smallvec::SmallVec<[T; N]> {
    fn visit_with<V: TypeVisitor<I>>(&self, visitor: &mut V) -> ControlFlow<V::BreakTy> {
        self.iter().try_for_each(|t| t.visit_with(visitor))
    }
}
```

In a few places, types specifically ask for `Vec` after pulling fields out the `InstantiatedPredicates` - in these cases, I just used `.to_vec()` to paper over the difference.

---

This will very likely be perf neutral for compile times (or even negative), but might be a small win with a relatively small change, so I figure it's worth submitting to benchmarking.

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 1]
```diff
@@ -132,7 +132,11 @@ fn with_fresh_ty_vars<'cx, 'tcx>(
         impl_def_id,
         self_ty: tcx.type_of(impl_def_id).subst(tcx, impl_substs),
         trait_ref: tcx.impl_trait_ref(impl_def_id).map(|i| i.subst(tcx, impl_substs)),
-        predicates: tcx.predicates_of(impl_def_id).instantiate(tcx, impl_substs).predicates,
+        predicates: tcx
+            .predicates_of(impl_def_id)
+            .instantiate(tcx, impl_substs)
+            .predicates
+            .to_vec(),
```
Comment (klensy) [inline compiler/rustc_trait_selection/src/traits/coherence.rs:139]: Here and lower: `to_vec` clones into new vec, maybe `into_vec` should be used?

Comment (Nathan-Fenner) [reply compiler/rustc_trait_selection/src/traits/coherence.rs:139]: Ah yes, in these cases `.into_vec()` should definitely be preferred.

**Comment by Nathan-Fenner** [comment]
I had mistakenly called `.to_vec()` when `.into_vec()` would be preferred, to avoid copying unnecessarily in the case where the predicates list was large. With this fix, the regressions _may_ disappear.

If there's still no improvement, we can try reducing the smallvec size to 3 or 2, and if neither of those work, close this as non-productive.

**Comment by Nathan-Fenner** [comment]
We're now deep in the weeds of hopeful empiricism; I've tried reducing the size from 4 to 3 to see if it makes any improvement.

**Comment by compiler-errors** [comment]
I'm happy to start a new perf run, but if this doesn't come back with any results I would probably say it's not worth testing more combinations without better data or another compelling reason to do so.

@bors try @rust-timer queue

**Comment by compiler-errors** [comment]
Thanks @Nathan-Fenner, but it doesn't look like this code is hot enough to be worth optimizing.

**Pull request closed without merging**